core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStoreMeta.java

```diff
@@ -0,0 +1,198 @@
+    mCapacityBytesOnDirs =
+        dirs.stream().map(PageStoreDir::getCapacityBytes).collect(Collectors.toList());
+    mUsedBytesOnDirs =
+        dirs.stream().map(PageStoreDir::getCachedBytes).collect(Collectors.toList());
+    mCapacity = mCapacityBytesOnDirs.stream().reduce(0L, Long::sum);
+    mUsed = mUsedBytesOnDirs.stream().reduce(0L, Long::sum);
+
+    if (shouldIncludeBlockIds) {
+      HashMultimap<PageStoreDir, String> dirToBlocksMap = HashMultimap.create();
+      try (LockResource lock = new LockResource(metaStore.getLock().readLock())) {
+        final Iterator<PageId> pages = metaStore.getPagesIterator();
+        while (pages.hasNext()) {
+          PageId pageId = pages.next();
+          try {
+            PageInfo pageInfo = metaStore.getPageInfo(pageId);
+            String blockId = pageId.getFileId();
+            dirToBlocksMap.put(pageInfo.getLocalCacheDir(), blockId);
+          } catch (PageNotFoundException e) {
+            // unreachable as we have acquired a read lock, and the page IDs are obtained from
+            // the iterator given by the metastore
+            throw new IllegalStateException(
+                String.format("Unreachable state reached because the page %s which "
+                    + "should be present in the page metastore cannot be found", pageId), e);
+          }
+        }
+      }
+      Map<PageStoreDir, Set<String>> blockLocations = Multimaps.asMap(dirToBlocksMap);
+      ImmutableMap.Builder<BlockStoreLocation, List<Long>> mapBuilder = ImmutableMap.builder();
+      ImmutableList.Builder<Long> listBuilder = ImmutableList.builder();
+      for (int i = 0; i < dirs.size(); i++) {
+        BlockStoreLocation location = new BlockStoreLocation(
+            DEFAULT_TIER, i, DEFAULT_MEDIUM);
+        Set<String> blocks = blockLocations.get(dirs.get(i));
+        if (blocks == null) {
+          mapBuilder.put(location, ImmutableList.of());
+          continue;
+        }
+        List<Long> blockIds = blocks.stream().map(Long::parseLong).collect(Collectors.toList());
+        mapBuilder.put(location, blockIds);
+        listBuilder.addAll(blockIds);
+      }
+      mBlockLocations = mapBuilder.build();
+      mBlocks = listBuilder.build();
+      mNumBlocks = mBlocks.size();
+    } else {
+      mBlockLocations = null;
+      mBlocks = null;
+      try (LockResource lock = new LockResource(metaStore.getLock().readLock())) {
+        final Iterator<PageId> pages = metaStore.getPagesIterator();
+        mNumBlocks = (int) Streams.stream(pages).map(PageId::getFileId).distinct().count();
+      }
+    }
+  }
+
+  @Nullable
+  @Override
+  public Map<String, List<Long>> getBlockList() {
+    if (mBlocks != null) {
+      return ImmutableMap.of(DEFAULT_TIER, mBlocks);
+    }
+    return null;
+  }
+
+  @Nullable
+  @Override
+  public Map<BlockStoreLocation, List<Long>> getBlockListByStorageLocation() {
+    return mBlockLocations;
+  }
```

Comment:
out of the scope of this pr,  we could change the interface to return an Optional instead.